Cloneable/CloneableGenerator.cs

```diff
@@ -41,77 +76,131 @@ private void GenerateCloneMethods(GeneratorExecutionContext context)
             foreach (var classSymbol in classSymbols)
             {
-                context.AddSource($"{classSymbol.Name}_cloneable.cs", SourceText.From(CreateCloneableCode(classSymbol), Encoding.UTF8));
+                var attribute = classSymbol.GetAttributes().First(x => x.AttributeClass.Equals(cloneableAttribute));
+                var isExplicit = (bool?)attribute.NamedArguments.FirstOrDefault(e => e.Key.Equals("ExplicitDeclaration")).Value.Value ?? false;
+                context.AddSource($"{classSymbol.Name}_cloneable.cs", SourceText.From(CreateCloneableCode(classSymbol, isExplicit), Encoding.UTF8));
             }
         }
 
-        private string CreateCloneableCode(INamedTypeSymbol classSymbol)
+        private string CreateCloneableCode(INamedTypeSymbol classSymbol, bool isExplicit)
         {
             string namespaceName = classSymbol.ContainingNamespace.ToDisplayString();
-            string fieldAssignmentsCode = GenerateFieldAssignmentsCode(classSymbol);
+            var fieldAssignmentsCode = GenerateFieldAssignmentsCode(classSymbol, isExplicit, false);
+            var fieldAssignmentsCodeSafe = fieldAssignmentsCode.Select(x =>
+            {
+                if (x.EndsWith("?.Clone"))
+                    return x + "Safe(referenceChain)";
+                return x;
+            });
+            fieldAssignmentsCode = fieldAssignmentsCode.Select(x =>
+            {
+                if (x.EndsWith("?.Clone"))
+                    return x + "()";
+                return x;
+            });
+
+            return $@"using System.Collections.Generic;
 
-            return $@"namespace {namespaceName}
+namespace {namespaceName}
 {{
     {GetAccessModifier(classSymbol)} partial class {classSymbol.Name}
     {{
+        /// <summary>
+        /// Creates a copy of {classSymbol.Name} with NO circular reference checking. This method should be used if performance matters.
+        /// 
+        /// <exception cref=""StackOverflowException"">Will occur on any object that has circular references in the hierarchy.</exception>
+        /// </summary>
         public {classSymbol.Name} Clone()
         {{
             return new {classSymbol.Name}
             {{
-{fieldAssignmentsCode}
+{string.Join($",{Environment.NewLine}", fieldAssignmentsCode)}
             }};
         }}
+
+        /// <summary>
+        /// Creates a copy of {classSymbol.Name} with circular reference checking. If a circular reference was detected, only a reference of the leaf object is passed instead of cloning it.
+        /// </summary>
+        /// <param name=""referenceChain"">Should only be provided if specific objects should not be cloned but passed by reference instead.</param>
+        public {classSymbol.Name} CloneSafe(Stack<object> referenceChain = null)
+        {{
+            if(referenceChain?.Contains(this) == true) 
+                return this;
+            referenceChain ??= new Stack<object>();
+            referenceChain.Push(this);
```

Comment:
I don't get why you pushed this to stack and then popped it without any usage.

Reply:
It acually is used in line 140. The code generator passes this as parameter for deep cloning. See PR description at the very bottom.

Reply:
Yes now I got it. That makes sense.